# Collaborative Filtering

In [74]:
import pandas as pd
import numpy as np
from sklearn.metrics import pairwise_distances
from scipy.spatial.distance import cosine, correlation

In [75]:
#We have .data file. This is not .csv file.
columns = ["userid", "movieid", "rating", "timestamp"]
userMovieData= pd.read_csv('C:/Users/Lenovo/Desktop/IPY/MoviesDataSet/u.data', sep='\t',names= columns)
userMovieData.head(10)

userid  movieid  rating  timestamp
0     196      242       3  881250949
1     186      302       3  891717742
2      22      377       1  878887116
3     244       51       2  880606923
4     166      346       1  886397596
5     298      474       4  884182806
6     115      265       2  881171488
7     253      465       5  891628467
8     305      451       3  886324817
9       6       86       3  883603013

In [76]:
userMovieData.drop( "timestamp", inplace = True, axis = 1 )

In [77]:
userMovieData.shape

(100000, 3)

In [78]:
print(len(userMovieData))
print("Unique UserIds  ---------> ",len( userMovieData['userid'].unique()))
print("Unique MovieIds ---------> ",len( userMovieData['movieid'].unique()))

100000
Unique UserIds  --------->  943
Unique MovieIds --------->  1682


#### Let's assume in this world there exists 943 people and 1682 movies. 943 people have rated different 1682 movies, so generated lot of data -  100000 rows

In [79]:
userMovieData.head(10)

userid  movieid  rating
0     196      242       3
1     186      302       3
2      22      377       1
3     244       51       2
4     166      346       1
5     298      474       4
6     115      265       2
7     253      465       5
8     305      451       3
9       6       86       3

#### To start with creation of "User-User Similarity Martix" it means creating a "Pivot". Pivots can be implemented in multiple ways, ranging from Excel to Programming. We are choosing obviously Python way! NOTE: We are creating pivot using pivot() method. pivot_table() is used to implement pivot tables. 

##### Refer:  https://github.com/rahulvaish/MachineLearningPython/blob/DataWrangling/UnderstandingPivotTables-GBToilets.ipynb

In [80]:
userMovieRatingRelationship = userMovieData.pivot( index='userid', columns='movieid', values = "rating" )

In [81]:
userMovieRatingRelationship

movieid  1     2     3     4     5     6     7     8     9     10    ...   \
userid                                                               ...    
1         5.0   3.0   4.0   3.0   3.0   5.0   4.0   1.0   5.0   3.0  ...    
2         4.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   2.0  ...    
3         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
4         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
5         4.0   3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
6         4.0   NaN   NaN   NaN   NaN   NaN   2.0   4.0   4.0   NaN  ...    
7         NaN   NaN   NaN   5.0   NaN   NaN   5.0   5.0   5.0   4.0  ...    
8         NaN   NaN   NaN   NaN   NaN   NaN   3.0   NaN   NaN   NaN  ...    
9         NaN   NaN   NaN   NaN   NaN   5.0   4.0   NaN   NaN   NaN  ...    
10        4.0   NaN   NaN   4.0   NaN   NaN   4.0   NaN   4.0   NaN  ...    
11        NaN   NaN   NaN   NaN   NaN   NaN   NaN   4.0   5.0   NaN  ...    
12        NaN   NaN   NaN   5.0   NaN   NaN   NaN   NaN   NaN   NaN  ...    
13        3.0   3.0   NaN   5.0   1.0   NaN   2.0   4.0   3.0   NaN  ...    
14        NaN   NaN   NaN   NaN   NaN   NaN   5.0   NaN   4.0   NaN  ...    
15        1.0   NaN   NaN   NaN   NaN   NaN   1.0   NaN   4.0   NaN  ...    
16        5.0   NaN   NaN   5.0   NaN   NaN   5.0   5.0   5.0   NaN  ...    
17        4.0   NaN   NaN   NaN   NaN   NaN   4.0   NaN   3.0   NaN  ...    
18        5.0   NaN   NaN   3.0   NaN   5.0   NaN   5.0   5.0   NaN  ...    
19        NaN   NaN   NaN   4.0   NaN   NaN   NaN   5.0   NaN   NaN  ...    
20        3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
21        5.0   NaN   NaN   NaN   2.0   NaN   5.0   NaN   5.0   NaN  ...    
22        NaN   2.0   NaN   5.0   NaN   NaN   NaN   NaN   NaN   NaN  ...    
23        5.0   NaN   NaN   NaN   NaN   NaN   4.0   4.0   NaN   NaN  ...    
24        NaN   NaN   NaN   NaN   NaN   NaN   4.0   5.0   5.0   NaN  ...    
25        5.0   NaN   NaN   NaN   NaN   NaN   4.0   4.0   NaN   NaN  ...    
26        3.0   NaN   NaN   NaN   NaN   NaN   3.0   NaN   4.0   NaN  ...    
27        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   4.0   NaN  ...    
28        NaN   NaN   NaN   NaN   3.0   NaN   5.0   NaN   NaN   NaN  ...    
29        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
30        NaN   3.0   NaN   NaN   NaN   NaN   4.0   NaN   NaN   NaN  ...    
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...    
914       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
915       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
916       4.0   3.0   3.0   4.0   3.0   NaN   4.0   NaN   5.0   NaN  ...    
917       3.0   NaN   1.0   NaN   NaN   NaN   NaN   NaN   5.0   NaN  ...    
918       3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
919       4.0   NaN   NaN   1.0   4.0   NaN   3.0   NaN   5.0   NaN  ...    
920       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
921       3.0   NaN   NaN   NaN   NaN   NaN   NaN   3.0   NaN   NaN  ...    
922       5.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
923       3.0   NaN   4.0   NaN   NaN   NaN   NaN   NaN   4.0   NaN  ...    
924       5.0   3.0   NaN   NaN   NaN   4.0   4.0   NaN   4.0   NaN  ...    
925       NaN   NaN   NaN   NaN   4.0   NaN   NaN   NaN   NaN   NaN  ...    
926       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
927       5.0   NaN   NaN   NaN   NaN   NaN   3.0   4.0   NaN   NaN  ...    
928       NaN   NaN   NaN   NaN   NaN   NaN   NaN   5.0   5.0   NaN  ...    
929       3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
930       3.0   NaN   NaN   NaN   NaN   NaN   NaN   3.0   NaN   NaN  ...    
931       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
932       4.0   NaN   NaN   NaN   NaN   NaN   4.0   NaN   5.0   NaN  ... 

In [82]:
# 943 X 1682 Crosschecked from Cell No.5 !!

In [83]:
# We see a lot of 'NAN's Missing values. Replacing 'NAN's with 0.
# Why 0 makes sense ?
# If user A has not seen the movie B, for B the rating is missing or 0. Same is the situation where movie B has been watched but not rated.

In [84]:
userMovieRatingRelationship.fillna( 0, inplace = True )

In [85]:
userMovieRatingRelationship

movieid  1     2     3     4     5     6     7     8     9     10    ...   \
userid                                                               ...    
1         5.0   3.0   4.0   3.0   3.0   5.0   4.0   1.0   5.0   3.0  ...    
2         4.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   2.0  ...    
3         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
4         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
5         4.0   3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
6         4.0   0.0   0.0   0.0   0.0   0.0   2.0   4.0   4.0   0.0  ...    
7         0.0   0.0   0.0   5.0   0.0   0.0   5.0   5.0   5.0   4.0  ...    
8         0.0   0.0   0.0   0.0   0.0   0.0   3.0   0.0   0.0   0.0  ...    
9         0.0   0.0   0.0   0.0   0.0   5.0   4.0   0.0   0.0   0.0  ...    
10        4.0   0.0   0.0   4.0   0.0   0.0   4.0   0.0   4.0   0.0  ...    
11        0.0   0.0   0.0   0.0   0.0   0.0   0.0   4.0   5.0   0.0  ...    
12        0.0   0.0   0.0   5.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
13        3.0   3.0   0.0   5.0   1.0   0.0   2.0   4.0   3.0   0.0  ...    
14        0.0   0.0   0.0   0.0   0.0   0.0   5.0   0.0   4.0   0.0  ...    
15        1.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0   4.0   0.0  ...    
16        5.0   0.0   0.0   5.0   0.0   0.0   5.0   5.0   5.0   0.0  ...    
17        4.0   0.0   0.0   0.0   0.0   0.0   4.0   0.0   3.0   0.0  ...    
18        5.0   0.0   0.0   3.0   0.0   5.0   0.0   5.0   5.0   0.0  ...    
19        0.0   0.0   0.0   4.0   0.0   0.0   0.0   5.0   0.0   0.0  ...    
20        3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
21        5.0   0.0   0.0   0.0   2.0   0.0   5.0   0.0   5.0   0.0  ...    
22        0.0   2.0   0.0   5.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
23        5.0   0.0   0.0   0.0   0.0   0.0   4.0   4.0   0.0   0.0  ...    
24        0.0   0.0   0.0   0.0   0.0   0.0   4.0   5.0   5.0   0.0  ...    
25        5.0   0.0   0.0   0.0   0.0   0.0   4.0   4.0   0.0   0.0  ...    
26        3.0   0.0   0.0   0.0   0.0   0.0   3.0   0.0   4.0   0.0  ...    
27        0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   4.0   0.0  ...    
28        0.0   0.0   0.0   0.0   3.0   0.0   5.0   0.0   0.0   0.0  ...    
29        0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
30        0.0   3.0   0.0   0.0   0.0   0.0   4.0   0.0   0.0   0.0  ...    
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...    
914       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
915       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
916       4.0   3.0   3.0   4.0   3.0   0.0   4.0   0.0   5.0   0.0  ...    
917       3.0   0.0   1.0   0.0   0.0   0.0   0.0   0.0   5.0   0.0  ...    
918       3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
919       4.0   0.0   0.0   1.0   4.0   0.0   3.0   0.0   5.0   0.0  ...    
920       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
921       3.0   0.0   0.0   0.0   0.0   0.0   0.0   3.0   0.0   0.0  ...    
922       5.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
923       3.0   0.0   4.0   0.0   0.0   0.0   0.0   0.0   4.0   0.0  ...    
924       5.0   3.0   0.0   0.0   0.0   4.0   4.0   0.0   4.0   0.0  ...    
925       0.0   0.0   0.0   0.0   4.0   0.0   0.0   0.0   0.0   0.0  ...    
926       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
927       5.0   0.0   0.0   0.0   0.0   0.0   3.0   4.0   0.0   0.0  ...    
928       0.0   0.0   0.0   0.0   0.0   0.0   0.0   5.0   5.0   0.0  ...    
929       3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
930       3.0   0.0   0.0   0.0   0.0   0.0   0.0   3.0   0.0   0.0  ...    
931       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    
932       4.0   0.0   0.0   0.0   0.0   0.0   4.0   0.0   5.0   0.0  ... 

In [86]:
#To check a part/section of DataFrame-
#userMovieRatingRelationship.iloc[10:20, 20:30]

# USER-USER SIMILARITY:

#### For calculating similarity between user-user, many similarity coefficients can be calculated. Most widely used similarity coefficients are Euclidean, Cosine, Pearson Correlation etc.
#### We will use cosine distance here. Here we are insterested in similarity. That means higher the value more similar they are. 


# *NOTE: The function we are going to use ( pairwise_distances() )gives us the distance, which we have to deduct it from 1. Also the function does not returns DataFrame, we have to perform a conversion over it!

In [87]:
UserUserSimilarity = 1 - pairwise_distances( userMovieRatingRelationship.as_matrix(), metric="cosine" )

In [88]:
UserUserSimilarityDataframe = pd.DataFrame(UserUserSimilarity)

In [89]:
UserUserSimilarityDataframe.shape

(943, 943)

#### Now we have a dataframe "UserUserSimilarityDataframe" which is the matrix form/representation showing the relationship beween all users

In [90]:
UserUserSimilarityDataframe

0         1         2         3         4         5         6    \
0    1.000000  0.166931  0.047460  0.064358  0.378475  0.430239  0.440367   
1    0.166931  1.000000  0.110591  0.178121  0.072979  0.245843  0.107328   
2    0.047460  0.110591  1.000000  0.344151  0.021245  0.072415  0.066137   
3    0.064358  0.178121  0.344151  1.000000  0.031804  0.068044  0.091230   
4    0.378475  0.072979  0.021245  0.031804  1.000000  0.237286  0.373600   
5    0.430239  0.245843  0.072415  0.068044  0.237286  1.000000  0.489255   
6    0.440367  0.107328  0.066137  0.091230  0.373600  0.489255  1.000000   
7    0.319072  0.103344  0.083060  0.188060  0.248930  0.201369  0.284951   
8    0.078138  0.161048  0.061040  0.101284  0.056847  0.183951  0.145650   
9    0.376544  0.159862  0.065151  0.060859  0.201427  0.551713  0.487024   
10   0.311124  0.129257  0.075912  0.104182  0.321710  0.260765  0.374882   
11   0.314678  0.121268  0.067345  0.075324  0.130664  0.239636  0.270165   
12   0.439743  0.223661  0.198196  0.120306  0.345217  0.472032  0.560047   
13   0.332107  0.226758  0.042780  0.070670  0.244741  0.405890  0.326199   
14   0.181473  0.455361  0.129912  0.148397  0.096349  0.194365  0.114483   
15   0.403987  0.125330  0.058530  0.084951  0.287556  0.392809  0.436113   
16   0.215819  0.223667  0.031320  0.053895  0.084729  0.204279  0.096406   
17   0.473875  0.179218  0.033788  0.035582  0.289495  0.585087  0.519394   
18   0.109084  0.101374  0.104290  0.158678  0.116843  0.129775  0.152058   
19   0.241498  0.060499  0.050696  0.086565  0.240738  0.181222  0.215534   
20   0.198413  0.204445  0.165823  0.189713  0.176494  0.138845  0.254091   
21   0.381042  0.064443  0.034462  0.082280  0.455057  0.245841  0.367952   
22   0.446288  0.146500  0.030955  0.061922  0.409268  0.442834  0.414847   
23   0.285399  0.178323  0.071430  0.175470  0.137635  0.317282  0.299696   
24   0.335553  0.148638  0.037031  0.084112  0.329765  0.379899  0.339270   
25   0.252066  0.445594  0.163477  0.151322  0.152247  0.219451  0.138174   
26   0.129579  0.185077  0.021322  0.074295  0.080846  0.128199  0.102063   
27   0.302737  0.102455  0.068035  0.151293  0.322380  0.211204  0.337934   
28   0.129084  0.175115  0.153260  0.143465  0.052199  0.177428  0.136528   
29   0.181933  0.230078  0.108480  0.144516  0.164811  0.169750  0.153965   
..        ...       ...       ...       ...       ...       ...       ...   
913  0.057587  0.039270  0.000000  0.000000  0.038711  0.042198  0.066867   
914  0.047805  0.260726  0.358316  0.136295  0.000000  0.081268  0.065122   
915  0.569066  0.116597  0.046124  0.055061  0.342110  0.401596  0.491683   
916  0.140963  0.284455  0.044585  0.043939  0.074416  0.133830  0.100724   
917  0.232308  0.049835  0.006246  0.000000  0.168863  0.279572  0.274143   
918  0.338279  0.376975  0.264406  0.182571  0.185523  0.302408  0.257285   
919  0.058859  0.328677  0.434379  0.171236  0.000000  0.074047  0.055770   
920  0.276885  0.167278  0.094758  0.100374  0.258688  0.209187  0.247998   
921  0.383251  0.122291  0.055179  0.132542  0.423096  0.214373  0.375270   
922  0.158199  0.231709  0.164234  0.094344  0.159031  0.134903  0.136162   
923  0.302857  0.244595  0.069298  0.082147  0.211030  0.334277  0.353069   
924  0.079341  0.048471  0.184687  0.168100  0.063369  0.057455  0.116820   
925  0.083955  0.390525  0.286220  0.232185  0.000000  0.130961  0.070597   
926  0.257381  0.113863  0.052784  0.127523  0.289893  0.146745  0.271234   
927  0.234315  0.091148  0.070027  0.053364  0.085181  0.234484  0.161209   
928  0.280072  0.114631  0.036707  0.048234  0.233524  0.429847  0.341675   
929  0.279170  0.288602  0.023085  0.073335  0.143641  0.204852  0.172656   
930  0.146437  0.512473  0.176416  0.138921  0.068424  0.209038  0.106790   
931  0.354255  0.047879  0.000000  0.023274  0.324288  0.521209  0.543948   
932  0.490353  0.102243  0.036302  0.071446  0.373698  0.403706  0.484337   
933  0

# *NOTE: Total Number of Rows and Columns in the Dataframe were 943.  After the transformation, the count of rows and columns is still 943, but the indexing srarted from 0 insted of 1.

#### Now we have a dataframe - represented  as "User-User Similarity Matrix" with numbers everywhere, how to determine the similarity for a particular user, i.e, which user is similar to another user on the basis of his/her movie liking/disliking ?                                                                                                                       

#### To answer this question and understand, we will be using an inbuild method idxmax(). But the problem is that, using "UserUserSimilarityDataframe.idxmax()" results that every person is similar to him/her self (Hopefully you have noticed the '1' in the diagnols). To solve this we need to turn the diagnols = 0 and then calculate.

In [91]:
np.fill_diagonal(UserUserSimilarity, 0 )

#Same Step as done previously
UserUserSimilarityDataframe = pd.DataFrame( UserUserSimilarity )
#Same Step as done previously
UserUserSimilarityDataframe.shape


(943, 943)

In [92]:
#Same Step as done previously
UserUserSimilarityDataframe.columns

RangeIndex(start=0, stop=943, step=1)

In [93]:
UserUserSimilarityDataframe.idxmax(axis=1)

0      915
1      700
2      862
3      749
4      306
5      473
6      649
7      599
8      734
9      473
10     270
11     631
12     233
13     912
14     180
15     822
16     656
17     473
18     578
19     289
20     365
21     637
22     406
23     467
24     912
25     431
26     402
27     366
28     826
29     460
      ... 
913    903
914    283
915     93
916    767
917     84
918    591
919    352
920     41
921     55
922    792
923    693
924    254
925    170
926    108
927    391
928    349
929    767
930    700
931    311
932    652
933     84
934    890
935    485
936    412
937    140
938    717
939    912
940    688
941    453
942    681
Length: 943, dtype: int64

# *NOTE: As the indexing started from 0 instead of 1. Therefore, to compute conclusions from the above results 930   700 =>  We can say that User (930 +1) = 931  is similar to User (700 + 1) = 701


# ITEM-ITEM SIMILARITY:

In [94]:
movieUserRatingRelationship = userMovieData.pivot( index='movieid', columns='userid', values = "rating" )

In [95]:
movieUserRatingRelationship

userid   1    2    3    4    5    6    7    8    9    10  ...   934  935  936  \
movieid                                                   ...                   
1        5.0  4.0  NaN  NaN  4.0  4.0  NaN  NaN  NaN  4.0 ...   2.0  3.0  4.0   
2        3.0  NaN  NaN  NaN  3.0  NaN  NaN  NaN  NaN  NaN ...   4.0  NaN  NaN   
3        4.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  4.0   
4        3.0  NaN  NaN  NaN  NaN  NaN  5.0  NaN  NaN  4.0 ...   5.0  NaN  NaN   
5        3.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
6        5.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  5.0  NaN ...   NaN  NaN  5.0   
7        4.0  NaN  NaN  NaN  NaN  2.0  5.0  3.0  4.0  4.0 ...   NaN  NaN  4.0   
8        1.0  NaN  NaN  NaN  NaN  4.0  5.0  NaN  NaN  NaN ...   NaN  NaN  NaN   
9        5.0  NaN  NaN  NaN  NaN  4.0  5.0  NaN  NaN  4.0 ...   NaN  1.0  4.0   
10       3.0  2.0  NaN  NaN  NaN  NaN  4.0  NaN  NaN  NaN ...   NaN  NaN  NaN   
11       2.0  NaN  NaN  4.0  NaN  NaN  3.0  3.0  NaN  4.0 ...   NaN  NaN  NaN   
12       5.0  NaN  NaN  NaN  NaN  4.0  5.0  NaN  NaN  5.0 ...   NaN  NaN  NaN   
13       5.0  4.0  NaN  NaN  NaN  2.0  NaN  NaN  NaN  3.0 ...   5.0  NaN  4.0   
14       5.0  4.0  NaN  NaN  NaN  5.0  NaN  NaN  NaN  NaN ...   NaN  NaN  4.0   
15       5.0  NaN  NaN  NaN  NaN  3.0  NaN  NaN  NaN  NaN ...   NaN  5.0  NaN   
16       5.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  4.0 ...   NaN  NaN  4.0   
17       3.0  NaN  NaN  NaN  4.0  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
18       4.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
19       5.0  3.0  NaN  NaN  NaN  4.0  NaN  NaN  NaN  NaN ...   NaN  NaN  5.0   
20       4.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  5.0   
21       1.0  NaN  NaN  NaN  3.0  3.0  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
22       4.0  NaN  NaN  NaN  NaN  3.0  5.0  5.0  NaN  5.0 ...   NaN  NaN  NaN   
23       4.0  NaN  NaN  NaN  NaN  4.0  3.0  NaN  NaN  5.0 ...   NaN  NaN  NaN   
24       3.0  NaN  NaN  NaN  4.0  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  4.0   
25       4.0  4.0  NaN  NaN  3.0  NaN  3.0  NaN  NaN  NaN ...   4.0  NaN  4.0   
26       3.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
27       2.0  NaN  NaN  NaN  NaN  NaN  4.0  NaN  NaN  NaN ...   NaN  NaN  NaN   
28       4.0  NaN  NaN  NaN  NaN  2.0  5.0  NaN  NaN  NaN ...   NaN  NaN  NaN   
29       1.0  NaN  NaN  NaN  4.0  NaN  3.0  NaN  NaN  NaN ...   NaN  NaN  NaN   
30       3.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
...      ...  ...  ...  ...  ...  ...  ...  ...  ...  ... ...   ...  ...  ...   
1653     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1654     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1655     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1656     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1657     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1658     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1659     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1660     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1661     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1662     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1663     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1664     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1665     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1666     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1667     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1668     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN   
1669     NaN  NaN  NaN  NaN  Na

In [38]:
movieUserRatingRelationship.fillna( 0, inplace = True )

In [39]:
ItemItemSimilarity = 1 - pairwise_distances( movieUserRatingRelationship.as_matrix(), metric="correlation" )

In [40]:
ItemItemSimilarityDataframe = pd.DataFrame(ItemItemSimilarity)

In [41]:
ItemItemSimilarityDataframe.shape

(1682, 1682)

In [55]:
ItemItemSimilarityDataframe

0         1         2         3         4         5         6     \
0     1.000000  0.234595  0.193362  0.226213  0.128840  0.015113  0.347354   
1     0.234595  1.000000  0.190649  0.409044  0.240712  0.030062  0.220022   
2     0.193362  0.190649  1.000000  0.227849  0.141368  0.065347  0.258855   
3     0.226213  0.409044  0.227849  1.000000  0.237298  0.021878  0.295489   
4     0.128840  0.240712  0.141368  0.237298  1.000000 -0.008594  0.205289   
5     0.015113  0.030062  0.065347  0.021878 -0.008594  1.000000  0.054415   
6     0.347354  0.220022  0.258855  0.295489  0.205289  0.054415  1.000000   
7     0.254490  0.206020  0.078636  0.352800  0.145866  0.012330  0.190670   
8     0.209502  0.077894  0.146181  0.229922  0.142541  0.079619  0.286572   
9     0.104655  0.072906  0.079608  0.138220 -0.033746  0.166084  0.178505   
10    0.221811  0.362190  0.265441  0.473171  0.281260 -0.013182  0.391409   
11    0.171780  0.343212  0.207190  0.457358  0.272113  0.074459  0.364064   
12    0.198063  0.074405  0.172782  0.254039  0.084938  0.228730  0.287245   
13    0.089739 -0.032029  0.078041  0.090185 -0.024039  0.209423  0.090469   
14    0.339027  0.076977  0.138165  0.113703  0.134558  0.010064  0.149448   
15    0.129048  0.031636  0.119143  0.092877  0.059505  0.115907  0.117292   
16    0.108860  0.320204  0.349635  0.319291  0.332488  0.045678  0.293315   
17   -0.023226  0.085478  0.104404  0.065539  0.052929  0.228790  0.028663   
18    0.022113 -0.069434  0.032821  0.012338  0.013776  0.186435  0.102364   
19    0.080876 -0.001471  0.122217  0.104686 -0.015551  0.152543  0.168835   
20    0.212938  0.244859  0.124356  0.151499  0.121435  0.030429  0.167264   
21    0.256862  0.368449  0.170528  0.434904  0.200503 -0.008163  0.272336   
22    0.072484  0.147752  0.127550  0.312218  0.139066  0.121458  0.188074   
23    0.281745  0.338009  0.301930  0.278431  0.134909  0.056764  0.365520   
24    0.333410  0.146921  0.189111  0.238576  0.159652  0.036667  0.310012   
25    0.064470  0.122634  0.148284  0.300564  0.163444  0.114321  0.111280   
26    0.072615  0.358680  0.208946  0.305548  0.188386  0.024980  0.192020   
27    0.376461  0.371013  0.144225  0.384661  0.213164 -0.000004  0.229629   
28    0.204595  0.526570  0.176465  0.335032  0.250817  0.018161  0.171396   
29    0.028232  0.022024  0.075252  0.147436  0.076392  0.162039  0.083848   
...        ...       ...       ...       ...       ...       ...       ...   
1652 -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1653  0.050142 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1654  0.034179  0.071415  0.091421  0.046743  0.088618 -0.005159  0.039920   
1655  0.005893 -0.016191 -0.012697  0.056858 -0.012859 -0.006708  0.042721   
1656  0.018215  0.099371 -0.009764 -0.016619  0.088618 -0.005159  0.006942   
1657  0.015635 -0.020829 -0.016334 -0.027801 -0.016543 -0.008630  0.033067   
1658  0.050142 -0.012451  0.057693  0.067863 -0.009889 -0.005159  0.039920   
1659  0.050142 -0.012451  0.057693  0.067863 -0.009889 -0.005159  0.039920   
1660  0.018215  0.099371  0.091421 -0.016619 -0.009889 -0.005159  0.023431   
1661 -0.035996 -0.015103 -0.011844 -0.020158 -0.011995 -0.006258 -0.031582   
1662 -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1663  0.036156  0.056823  0.001141  0.041604  0.037974  0.110130  0.022343   
1664 -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1665 -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1666 -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1667 -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1668 -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1669 -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1670 -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1671  0.003186  0.041716 -0.0138

# *NOTE: The indexing before and after using pairwise_distances(). Before the indexing was starting from 1, and now, its starting from 0.

In [44]:
ItemItemSimilarityDataframe.idxmax(axis=1)

0          0
1          1
2          2
3          3
4          4
5          5
6          6
7          7
8          8
9          9
10        10
11        11
12        12
13        13
14        14
15        15
16        16
17        17
18        18
19        19
20        20
21        21
22        22
23        23
24        24
25        25
26        26
27        27
28        28
29        29
        ... 
1652    1652
1653    1653
1654    1654
1655    1655
1656    1656
1657    1657
1658    1658
1659    1658
1660    1660
1661    1661
1662    1662
1663    1663
1664    1662
1665    1662
1666    1662
1667    1662
1668    1662
1669    1662
1670    1670
1671    1671
1672    1672
1673    1673
1674    1674
1675    1674
1676    1676
1677    1677
1678    1677
1679    1677
1680    1680
1681    1681
Length: 1682, dtype: int64

# In USER-USER Similarity 
# X axis = USER | Y axis = ITEM ------> X axis = USER| Y axis = USER.


# In ITEM-ITEM Similarity 
# X axis = ITEM | Y axis = USER ------> X axis = ITEM| Y axis = ITEM.


#### Now, for ITEM-ITEM Similarity relationship, if we are also interested in decoding the item names, then we will have to use u.item data set. It contains ItemId (movie Id) and its corresponding Item Names (movie Name).

In [66]:
itemData= pd.read_csv('C:/Users/Lenovo/Desktop/IPY/MoviesDataSet/u.item', sep='\|', header = None )
itemData = itemData.iloc[:,:2]
itemData.columns = ['movieid', 'title']
itemData.head(10)

C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  """Entry point for launching an IPython kernel.


movieid                                              title
0        1                                   Toy Story (1995)
1        2                                   GoldenEye (1995)
2        3                                  Four Rooms (1995)
3        4                                  Get Shorty (1995)
4        5                                     Copycat (1995)
5        6  Shanghai Triad (Yao a yao yao dao waipo qiao) ...
6        7                              Twelve Monkeys (1995)
7        8                                        Babe (1995)
8        9                            Dead Man Walking (1995)
9       10                                 Richard III (1995)

In [67]:
#Introducing a new column 'similarity' in the original dataframe.
itemData['similarity'] = ItemItemSimilarityDataframe.iloc[0]

In [68]:
itemData.head(10)

movieid                                              title  similarity
0        1                                   Toy Story (1995)    1.000000
1        2                                   GoldenEye (1995)    0.234595
2        3                                  Four Rooms (1995)    0.193362
3        4                                  Get Shorty (1995)    0.226213
4        5                                     Copycat (1995)    0.128840
5        6  Shanghai Triad (Yao a yao yao dao waipo qiao) ...    0.015113
6        7                              Twelve Monkeys (1995)    0.347354
7        8                                        Babe (1995)    0.254490
8        9                            Dead Man Walking (1995)    0.209502
9       10                                 Richard III (1995)    0.104655

In [69]:
userMovieData.sort_values( ["similarity"], ascending = False )[1:10]

userid  movieid  rating  similarity
49      178      332       3    0.457677
120      94      789       4    0.454544
116     271      346       4    0.431789
150     299      955       4    0.423975
180     279     1240       1    0.422991
404     180      462       5    0.416770
94      145      275       2    0.407829
117      62       21       3    0.404908
221      58     1098       4    0.391073

In [70]:

def get_similar_movies( movieid, topN = 5 ):
  userMovieData['similarity'] = ItemItemSimilarityDataframe.iloc[movieid -1]
  top_n = userMovieData.sort_values( ["similarity"], ascending = False )[0:topN]
  print( "Similar Movies to: ", )
  return top_n

In [71]:
get_similar_movies( 118 )

Similar Movies to: 


userid  movieid  rating  similarity
117      62       21       3    1.000000
120      94      789       4    0.629867
404     180      462       5    0.542379
545     197      346       3    0.509549
596      13      864       4    0.489693